docs/architecture/adr-047-extend-upgrade-plan.md

```diff
@@ -0,0 +1,206 @@
+We will add the following validation:
+1.  If `UpgradeInstructions` are provided:
+    1.  There MUST be at least one entry in `artifacts`.
+    1.  All of the `artifacts` MUST have a unique `platform`.
+    1.  For each `Artifact`, if the `url` contains a `checksum` query parameter:
+        1. The `checksum` query parameter value MUST be in the format of `{checksum_algo}:{checksum}`.
+        1. The `{checksum}` from the query parameter MUST equal the `checksum` provided in the `Artifact`.
+        1. The `{checksum_algo}` from the query parameter MUST equal the `checksum_algo` provided in the `Artifact`.
+1.  The following validation is currently done using the `info` field. We will apply similar validation to the `UpgradeInstructions`.
+    For each `Artifact`:
+    1.  The `platform` MUST have the format `{OS}/{CPU}` or be `"any"`.
+    1.  The `url` field MUST NOT be empty.
+    1.  The `url` field MUST be a proper URL.
+    1.  A `checksum` MUST be provided either in the `checksum` field or as a query parameter in the `url`.
+    1.  If the `checksum` field has a value and the `url` also has a `checksum` query parameter, the two values MUST be equal.
+    1.  The `url` MUST return either a file or an archive containing either `bin/{DAEMON_NAME}` or `{DAEMON_NAME}`.
+    1.  If a `checksum` is provided (in the field or as a query param), the checksum of the result of the `url` MUST equal the provided checksum.
+
+Downloading of an `Artifact` will happen the same way that URLs from `info` are currently downloaded.
+
+### Cosmovisor Updates
+
+If the `upgrade-info.json` file does not contain any `UpgradeInstructions`, existing functionality will be maintained.
+
+We will update Cosmovisor to look for and handle the new `UpgradeInstructions` in `upgrade-info.json`.
+If the `UpgradeInstructions` are provided, we will do the following:
+1.  The `info` field will be ignored.
+1.  The `artifacts` field will be used to identify the artifact to download based on the `platform` that Cosmovisor is running in.
+1.  If a `checksum` is provided (either in the field or as a query param in the `url`), and the downloaded artifact has a different checksum, the upgrade process will be interrupted and Cosmovisor will exit with an error.
+1.  If a `pre_run` command is defined, it will be executed at the same point in the process where the `app pre-upgrade` command would have been executed.
+    It will be executed using the same environment as other commands run by Cosmovisor.
+1.  If a `post_run` command is defined, it will be executed after executing the command that restarts the chain.
+    It will be executed in a background process using the same environment as the other commands.
+    Any output generated by the command will be logged.
+    Once complete, the exit code will be logged.
+
+We will update Cosmovisor to allow JSON of the new `UpgradeInstructions` to be allowed in the existing `info` field.
```

Comment:
why? I think we shouldn't mix the two any more. They should co-exist:
```
{
  "info": "some text"
  "instructions: { object }
}
```

Reply:
Being able to retroactively change an upgrade plan is sometimes desirable. Allowing a url with `instructions` JSON to be provided allows such a change. This is important in instances when security vulnerabilities are found and fixed and you don't want nodes to ever try using the vulnerable version.

I've updated the sentence to:
> We will update Cosmovisor to allow the `info` field to be a URL to JSON of the new `UpgradeInstructions` structure.

Reply:
Another option that would handle this situation very well is to add a `signer` next to the `checksum`. If a `signer` is provided, then the asset must be signed by that signer. It's not expected that someone will want to use a `checksum` and `signer` but I don't think we should disallow it.

By having a `signer`, it would be possible to securely change an asset without having to issue a new upgrade proposal.

Reply:
But why putting this in `info`? I wouldn't touch it (except deprecating it's meaning). Maybe we can create a new attribute: `Plan.InstructionsURL`. And we should make a validation that both `Plan.instructions` or `Plan.instructionsURL` can't be defined at the same time.

Reply:
I held off on adding a `Plan.InstructionsURL` field since that functionality already exists on the info field. All that's changing is the addition of a new structure option. Additionally, we'd want to also allow definition of a `checksum` and/or `signer` (like the other asset urls). The `Asset` message could be used here except that a `platform` isn't desirable in this case. Comments could be used to indicate that though. But the security available for the other assets should also be available for the instructions url.

For this type of thing to work securely, we really need to add a `signer` field to the `Asset` message. However, I know that that adds a considerable amount of work to implement. I thought it might be best to have discussions on it first.

Reply:
I resolved this discussion: https://github.com/cosmos/cosmos-sdk/pull/10602#discussion_r789032275 in order to continue it here.

After discussing this with the group, we decided to not do a `signer` field and to also deprecate the url/json in the `info` field. We also decided not to include an `InstructionsURL` field. I will update the doc to reflect this, then resolve this discussion.

### Reasoning

The chain only ever has at most one upgrade plan. If a chain has a version that shouldn't be used (e.g. security vulnerability), they would put in a new upgrade plan with a fixed version. If the bad version isn't yet in use, the new plan will overwrite the old plan and the bad version will be ignored. If the bad version is already running, then the new upgrade plan will update the nodes as usual.

The functionality of a `signer` is needed when an asset needs to change after the upgrade plan has been submitted. If the instructions can no longer be an asset, the only things that would be changing are the executables (or archives with the executables). If those change, though, then a new upgrade plan should be submitted anyway.